docker-compose/docker-compose.yml

```diff
@@ -0,0 +1,82 @@
+version: '3.4'
+
+networks:
+  p2pool:
+    driver: bridge
+
+volumes:
+  p2pool: 
+    name: ${P2POOL_VOLUME}
+  monero:
+    name: ${MONERO_VOLUME}
+
+services:
+  monero:
+    image: monero:latest
+    build:
+      context: monero
+      args:
+        - MONERO_GIT_TAG=$MONERO_GIT_TAG
+    container_name: p2pool-monero
+    networks:
+      - p2pool
+    ports:
+      - 18080:18080/tcp
+    volumes:
+      - monero:/home/monero/.bitmonero:rw
+      - /dev/null:/home/monero/.bitmonero/bitmonero.log:rw
+    restart: unless-stopped
+    command: >-
+      --zmq-pub tcp://0.0.0.0:18083
+      --disable-dns-checkpoints
+      --enable-dns-blocklist
+      --non-interactive
+      --p2p-bind-ip=0.0.0.0
+      --p2p-bind-port=18080
+      --rpc-bind-ip=0.0.0.0
+      --rpc-bind-port=18081
+      --confirm-external-bind
+      ${PRUNE_NODE}
+
+  p2pool:
+    image: p2pool:latest
+    build: ../
+    container_name: p2pool-p2pool
+    networks:
+      - p2pool
+    ports:
+      - ${P2POOL_STRATUM_PORT}:3333/tcp
+      - 37889:37889/tcp
+    volumes:
+      - p2pool:/home/p2pool/.p2pool:rw
+      - /dev/null:/home/p2pool/.p2pool/p2pool.log:rw
+    depends_on:
+      - monero
+    restart: unless-stopped
+    command: >-
+      --host monero
+      --wallet ${WALLET_ADDRESS}
+      --loglevel ${P2POOL_LOGLEVEL}
+
+  xmrig:
+    image: xmrig:latest
+    build: xmrig
+    container_name: p2pool-xmrig
+    networks:
+      - p2pool
+    privileged: true
+    cap_add:
+      - ALL
+    volumes:
+      - /dev:/dev:ro
+      - /lib/modules:/lib/modules:ro
+      - /dev/hugepages:/dev/hugepages:rw
+    depends_on:
+      - p2pool
+    restart: unless-stopped
+    command: >-
+      ${FIXED_MINING_DIFFICULTY}
+      ${NUM_MINING_CPU}
+      --randomx-1gb-pages
+      -o p2pool:3333
+
```

Comment:
Does MSR mod work when xmrig is run in Docker container? IIRC it doesn't work in a VM.

Reply:
docker isnt a vm, no virtualization is done (so performance is not impacted either).

Yes MSR works, but does require some kernel configuration first (not specific to docker, same config is needed with or without docker)

```p2pool-xmrig | [2021-11-11 15:19:43.844]  msr      register values for "intel" preset have been set successfully (1 ms)```


Full log:  (I just now notice it does allocate "huge pages" but not "1GB pages" so I'll look into that)

```
p2pool-xmrig |  * ABOUT        XMRig/6.15.3 gcc/9.3.0
p2pool-xmrig |  * LIBS         libuv/1.34.2 OpenSSL/1.1.1f hwloc/2.1.0
p2pool-xmrig |  * HUGE PAGES   supported
p2pool-xmrig |  * 1GB PAGES    supported
p2pool-xmrig |  * CPU          Intel(R) Core(TM) i9-7940X CPU @ 3.10GHz (1) 64-bit AES
p2pool-xmrig |                 L2:14.0 MB L3:19.2 MB 14C/28T NUMA:1
p2pool-xmrig |  * MEMORY       59.7/62.5 GB (96%)
p2pool-xmrig |                 DIMM_A1: 8 GB DDR4 @ 2666 MHz CMK32GX4M4A2666C16  
p2pool-xmrig |                 DIMM_A2: 8 GB DDR4 @ 2666 MHz CMK32GX4M4A2666C16  
p2pool-xmrig |                 DIMM_B1: 8 GB DDR4 @ 2666 MHz CMK32GX4M4A2666C16  
p2pool-xmrig |                 DIMM_B2: 8 GB DDR4 @ 2666 MHz CMK32GX4M4A2666C16  
p2pool-xmrig |                 DIMM_C1: 8 GB DDR4 @ 2666 MHz CMK32GX4M4A2666C16  
p2pool-xmrig |                 DIMM_C2: 8 GB DDR4 @ 2666 MHz CMK32GX4M4A2666C16  
p2pool-xmrig |                 DIMM_D1: 8 GB DDR4 @ 2666 MHz CMK32GX4M4A2666C16  
p2pool-xmrig |                 DIMM_D2: 8 GB DDR4 @ 2666 MHz CMK32GX4M4A2666C16  
p2pool-xmrig |  * MOTHERBOARD  ASUSTeK COMPUTER INC. - ROG STRIX X299-E GAMING
p2pool-xmrig |  * DONATE       1%
p2pool-xmrig |  * ASSEMBLY     auto:intel
p2pool-xmrig |  * POOL #1      p2pool:3333 algo auto
p2pool-xmrig |  * COMMANDS     hashrate, pause, resume, results, connection
p2pool-xmrig |  * OPENCL       disabled
p2pool-xmrig |  * CUDA         disabled
p2pool-xmrig | [2021-11-11 15:19:43.843]  net      use pool p2pool:3333  172.26.0.3
p2pool-xmrig | [2021-11-11 15:19:43.843]  net      new job from p2pool:3333 diff 100001 algo rx/0 height 2490996 (1 tx)
p2pool-xmrig | [2021-11-11 15:19:43.843]  cpu      use argon2 implementation AVX-512F
p2pool-xmrig | [2021-11-11 15:19:43.844]  msr      register values for "intel" preset have been set successfully (1 ms)
p2pool-xmrig | [2021-11-11 15:19:43.844]  randomx  init dataset algo rx/0 (28 threads) seed c7743ee59c670469...
p2pool-xmrig | [2021-11-11 15:19:43.981]  randomx  failed to allocate RandomX dataset using 1GB pages
p2pool-xmrig | [2021-11-11 15:19:43.981]  randomx  allocated 2336 MB (2080+256) huge pages 89% 1040/1168 +JIT (137 ms)
p2pool-monero | 2021-11-11 15:19:45.258	E Transaction not found in pool
p2pool-monero | 2021-11-11 15:19:45.866	E Transaction not found in pool
p2pool-xmrig | [2021-11-11 15:19:45.881]  randomx  dataset ready (1900 ms)
p2pool-xmrig | [2021-11-11 15:19:45.881]  cpu      use profile  rx  (14 threads) scratchpad 2048 KB
p2pool-xmrig | [2021-11-11 15:19:45.918]  cpu      READY threads 14/14 (14) huge pages 100% 14/14 memory 28672 KB (37 ms)
```

Reply:
The failure to allocate 1GB huge pages was due to my kernel not supporting them.  I configured a pool of 1G huge pages in addition to the 2M huge pages and now xmrig is happy.

```
$ hugeadm --pool-list
      Size  Minimum  Current  Maximum  Default
   2097152     1280     1280     1280        *
1073741824       16       16       16         
```